### DISCUSSION TITLE
Using `@mdx-js/react` 2.0.0-rc1 inside Next.js

### DISCUSSION THREAD

**Discussion by iAdramelk** [mdx-js/mdx#1766] [Q&A] [answered]
I'm trying to update from `2.0.0-next.9` to `2.0.0-rc1`, it mostly works and fixes bunch of bugs, but I have a problem with making `@mdx-js/react` work.

So my setup is something like this. I have mdx file (simplified version):

```js
import Layout from "layouts/LayoutName";

export const meta = { key: "value" };

export default function Wrapper (props) {
  return <Layout {...props} meta={meta} />;
};
```

And I have layout file (simplified version):

```js
import { MDXProvider } from "@mdx-js/react";

const components = {
  h1: (props) => { return <h1 {...props}  /> }
};

export default function LayoutName ({ children }) {
  return <MDXProvider components={components}>{children}</MDXProvider>;
}
```

This code is running on `Next.js` with both client and server rendering modes. It works without problem on `2.0.0-next.9`, but on `2.0.0-rc1` it fails with error:

```
Error [ERR_REQUIRE_ESM]: require() of ES Module /node_modules/@mdx-js/react/index.js from /filename.js not supported.
Instead change the require of index.js in filename.js to a dynamic import() which is available in all CommonJS modules.
```

The problem here from that I understand is that Nextjs didn't compile `@mdx-js/react` to the server and leave it as require instead. Lines from created bundle file:

```
/***/ "@mdx-js/react":
/*!********************************!*\
  !*** external "@mdx-js/react" ***!
  \********************************/
/***/ (function(module) {

"use strict";
module.exports = require("@mdx-js/react");;

/***/ }),
```

Without `MDXProvider` everything works as intended.

And I can't find a way how I can replace it with `import` only for the SSR build. Is there a known workaround for this problem?

I can probably try to manually push `components` prop to children, but it looks hacky.

**Comment by ChristianMurphy** [comment, marked as answer]
@iAdramelk if you are using Next.js 11.1 or higher, you may need to enable https://nextjs.org/blog/next-11-1#es-modules-support

**Reply by ChristianMurphy** [reply]
This Gist include points for both Next.js and other build tools https://gist.github.com/sindresorhus/a39789f98801d908bbc7ff3ecc99d99c

**Reply by iAdramelk** [reply]
Thanks for you suggestion, I tried to update Next to 11.1.2 and enabled es-modules support. It started loading es-modules without throwing error, but while `MDXProvider` is now loaded it is for some reason ignores `components` prop and rendering stuff without it. Debugging it for now.

**Reply by wooorm** [reply]
You probably did not configure the webpack loader to support it.
did you find the migration docs? https://v2.mdxjs.com/migrating/v2/#mdx-jsloader

**Reply by iAdramelk** [reply]
Yep, missed it. Thank you. Will try.

**Reply by iAdramelk** [reply]
Ok, I added `providerImportSource: '@mdx-js/react'` to loader settings, but it still does not work for my case. My guess from studying generated code:

```
/*@jsxRuntime automatic @jsxImportSource react*/
const {Fragment: _Fragment, jsx: _jsx} = arguments[0];
function MDXContent(props = {}) {
  const _components = Object.assign({}, props.components), {wrapper: MDXLayout} = _components;
  const _content = _jsx(_Fragment, {});
  return MDXLayout ? _jsx(MDXLayout, Object.assign({}, props, {
    children: _content
  })) : _content;
}
return {
  default: MDXContent
};
```

Is that because I use `MDXProvider` inside `MDXLayout` (default exported component) they didn't apply to `MDXContent` and I can't wrap `mdx` component itself in `MDXProvider` or pass `components` to it because I use mdx and mdx-loader to render full pages. Is there some workaround for it or maybe I'm doing something wrong?

Update: Correction. Technically I can move `MDXProvider` to `_app.js` in next and it should start working then, but idea here is to have separate components list for each layout to reduce page weight and make it possible to have different header styles for example.

**Reply by wooorm** [reply]
Where are you seeing that generated code? That is not code that `@mdx-js/loader` generates.

**Reply by wooorm** [reply]
Layouts get rendered children. They wrap stuff. They can’t then change the rendered children again.

**Reply by iAdramelk** [reply]
Generated code above is from playground on mdx site, it didn't contained imports, but I thought that it would be enough to show my point. Anyway, real code from the loader looks like this:

```
/*@jsxRuntime automatic @jsxImportSource react*/
import {Fragment as _Fragment, jsx as _jsx} from "react/jsx-runtime";
import {useMDXComponents as _provideComponents} from "@mdx-js/react";
import Layout from "layouts/SitePage";
export const meta = {};
const MDXLayout = function Wrapper(props) {
  return _jsx(Layout, Object.assign({}, props, {
    meta: meta
  }));
};
;
function MDXContent(props = {}) {
  const _components = Object.assign({
    p: "p"
  }, _provideComponents(), props.components);
  const _content = _jsx(_Fragment, {
    children: _jsx(_components.p, {
      children: "Example text."
    })
  });
  return MDXLayout ? _jsx(MDXLayout, Object.assign({}, props, {
    children: _content
  })) : _content;
}
export default MDXContent;
```

It's a little different, but I think its still impossible to pass components here from `MDXLayout`.

In `2.0.0-next.9` `MDXProvider` inside layouts worked (but maybe it was a bug and not a feature). If it not-working now is an expected behavior, when what is the best way to style stand-alone `mdx` pages which you do not explicitly import by hand? 

I want to be able to be able to write mdx file, add default export for layout (I'm using a plugin that converts `layout` field from frontmatter to default export to make it easier for editors), render it to a page with mdx loader and then see page content wrapper in layout with components styled in the way apropriative to this layout. Styling html with css is not enough, I need to do stuff like adding anchor links to headers, add copy button to code blocks, etc.

What is the best way to do it now?

**Reply by wooorm** [reply]
Can you share information about what components you want to pass and why some are in certain files and not others?

**Reply by iAdramelk** [reply]
Ok. So I'm working on the static site. Site have few hundreds of docs pages, fifty or so blog posts, and some general use pages.

Right now I'm using `remark` plugin that I wrote to switch layouts for them using frontmatter. E. g.:

```
---
layout: blog
---

Blog post here.
```

Plugin will add all required imports and exports to the mdx file and will pass all frontmatter fields which are not layout to the component as `meta` prop. As a result each page will be rendered in it's own layout.

Why I want to use different components for them:

1. We have different design for text, headers and lists on blog and docs pages (blog post styles generally use larger font sizes and different colors at times).
2. We have different behavior for some components in docs and in the blog posts. E. g. headers in docs will have anchor icon shown near them on hover like github does, blog posts does not need to do it.
3. We have some components that are often used in docs like `Tabs` or `Admonition` so it make sense to add them to `docs`'s component list instead of importing them by hand every time, but we do not need them on blog post page and I don't want to increase blog page's bundle size with components I don't plan to use.

**Reply by wooorm** [reply]
I wonder if generating this would work:

```js
function MDXContent(props = {}) {
  return MDXLayout
    ? _jsx(MDXLayout, Object.assign({}, props, {children: _jsx(_create)}))
    : _create();
  function _create() {
    const _components = Object.assign({p: "p"}, _provideComponents(), props.components);
    return _jsx(_Fragment, {
      children: _jsx(_components.p, {/* ... */})
    });
  }
}
```

**Reply by iAdramelk** [reply]
If `_provideComponents` takes components from `context` it may work this way, but I may be mistaken here.

Perfect solution for m would be a way to set `components` without using `MDXProvider` at all, but I don't have an idea how it can be possible to implement.

**Reply by wooorm** [reply]
Can you explain why that would be “perfect”? I think that’s a bit different from this case?

> [...] would be a way to set components without using MDXProvider at all

You could pass components to the MDX content component perhaps? MDXProvider is only really needed for MDX files inside other MDX files: https://v2.mdxjs.com/docs/using-mdx/#mdx-provider.

**Reply by wooorm** [reply]
Confirmed that this output javascript does work. For:

```mdx
import { MDXProvider } from '@mdx-js/react'

export default function Layout({children}) {
  return <MDXProvider components={{h1: 'h2'}}>{children}</MDXProvider>
}


# hi
```

Turning it into:

```js
import {useMDXComponents as _provideComponents} from "@mdx-js/react";
import {MDXProvider} from '@mdx-js/react';
const MDXLayout = function Layout({children}) {
  return <MDXProvider components={{
    h1: 'h2'
  }}>{children}</MDXProvider>;
};
function MDXContent(props = {}) {
  return MDXLayout ? <MDXLayout {...props}><_createMdxContent /></MDXLayout> : _createMdxContent();
  function _createMdxContent() {
    const _components = Object.assign({h1: "h1"}, _provideComponents(), props.components);
    return <><_components.h1>{"hi"}</_components.h1></>;
  }
}
export default MDXContent;
```

**Reply by iAdramelk** [reply]
Well, my point was that I'd prefer not to use `MDXProvider` if possible, to increase performance and reduce bundle size.

My use case is the docs site with hundreds of mdx pages so I can't wrap each of them inside react component and pass props manually, so I need to either pass them from the layout or from mdx file itself. Right now I use `MDXProvider` inside layout because there are no other way to pass components to the page. But if was possible to pass component list in some other way I would use it instead.

For example if I can add named export inside mdx file that loads and renders components I would then just write a plugin to add something like `export const components = {}` or `export { default as components } from "path-to-components";` to each mdx file at build time. 

Or if I can have access to `MDXContent` inside layout instead of `children`, I would do something like that:

```
function Layout ({ MdxContent }) {
  return <MDXContent  components={compoentns} />
}
```

But `MDXProvider` will do if there are no better options.

**Reply by wooorm** [reply]
I solved this on xdm just now. It can be added to mdx-js/mdx in the next rc/the actual release. See: https://github.com/wooorm/xdm/releases/tag/3.1.0. I want to be a bit conservative and slow with MDX RC releases.

> Or if I can have access to MDXContent inside layout instead of children, I would do something like that:

So, I’m guessing that *this* is the problem. Next doesn’t give you that, apparently? Maybe that can be fixed by them. Or Next already has that. This is unrelated to MDX — it happens on the Next level.

Regardless, why not this?

```js
// pages/index.js
import {MDXProvider} from '@mdx-js/react'

// ...

export default function Post(props) {
  const components = someCondition ? specialComponents : otherComponents
  return (
    <MDXProvider components={components}>
      <main {...props} />
    </MDXProvider>
  )
}
```

**Reply by iAdramelk** [reply]
Thanks for the fix, I really appreciate it!

```
// pages/index.js
import {MDXProvider} from '@mdx-js/react'

// ...

export default function Post(props) {
  const components = someCondition ? specialComponents : otherComponents
  return (
    <MDXProvider components={components}>
      <main {...props} />
    </MDXProvider>
  )
}
```

This is pretty much that I'm doing with `2.0.0-next.9` right now and it works. But it still includes `MDXProvider` which adds additional component to the bundle and creates React.context that somewhat slows page render. It's probably not super bad all things considered, but if I could add `components` to the mdx component directly without using context it would probably be more efficient.

**Reply by wooorm** [reply]
Yeah this isn't about mdx. MDX exposes components. It let's you do exactly that. Calling them is what next does. Maybe there are options for that there.

**Reply by iAdramelk** [reply]
> So, I’m guessing that this is the problem. Next doesn’t give you that, apparently? Maybe that can be fixed by them. Or Next already has that. This is unrelated to MDX — it happens on the Next level.

Not exactly. Next didn't know anything about mdx and don't have special integration with it (at least I didn't use it). Its just loads react component with a default export as a page if it is placed inside `pages` folder. 

So I'm using `@mdx-js/loader` to compile mdx to the javascript and then Next loads it as a page without any changes. What I want to do is to use just `@mdx-js/loader` to compile static js file from mdx with layout and components ans save it as `.js` file to the filesystem. Same as Docsaurus is doing without Next. I think Gatsby is doing more or less the same thing, but I haven't check their implementation.

Theoretically I can do my own webpack loader for MDX which will just wrap generated `MDXContent` with something else based on the filename, but it will be cool to have it working by default.

**Reply by wooorm** [reply]
> Its just loads react component with a default export as a page if it is placed inside pages folder.

Again, next loads it *and renders it*. We don't do that. We export a component. We do exactly what you describe you want.

**Reply by iAdramelk** [reply]
I will check it out. Maybe I understand something wrong. Thanks anyway.

**Comment by wooorm** [comment]
~nvm~
